### ISSUE TITLE
gen_tcp:connect(loopback, ...) attempts to connect to host "loopback" rather than the loopback address

### ISSUE THREAD

**Issue by mxxk** [erlang/otp#5044]
**Describe the bug**
`gen_tcp:connect(loopback, ...)` interprets the atom `'loopback'` as a hostname rather than the loopback address.

**To Reproduce**

1. Start a local TCP server: `nc -vl localhost 4321`
2. Try to connect to it from the Erlang shell: 
    ```
    1> gen_tcp:connect(loopback, 4321).
    {error,nxdomain}
    ```

**Expected behavior**
`gen_tcp:connect(loopback, ...)` connects to the loopback address.

**Affected versions**
Erlang/OTP 24 [erts-12.0.3] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:1]

**Additional context**
- If this is not a bug, it may be helpful to call out this behavior in the `gen_tcp:connect/N` docs.

- It seems that the `Address` in `gen_tcp:connect/N` has an overlapping type specification for the atoms `'any'` and `'loopback'`, as they are valid for both `socket_address()` and `hostname()`:

    https://github.com/erlang/otp/blob/645ea8110da97687de302791829b4a6119a9b396/lib/kernel/src/gen_tcp.erl#L154-L156

    https://github.com/erlang/otp/blob/645ea8110da97687de302791829b4a6119a9b396/lib/kernel/src/inet.erl#L185-L186

    https://github.com/erlang/otp/blob/645ea8110da97687de302791829b4a6119a9b396/lib/kernel/src/inet.erl#L105

    Perhaps this is related to why `'loopback'` is treated as a hostname...

- Interestingly, `gen_udp:send/N` does not share the same overlap, since it has distinct forms for `Host, Port` and `Destination`:

    https://github.com/erlang/otp/blob/645ea8110da97687de302791829b4a6119a9b396/lib/kernel/src/gen_udp.erl#L130-L131

    https://github.com/erlang/otp/blob/645ea8110da97687de302791829b4a6119a9b396/lib/kernel/src/gen_udp.erl#L140

- A reproduction of the TCP example, as well as a comparison to the UDP example, can be found at https://gist.github.com/mxxk/a85fc8454803f8fd3e1df7057beec987

**Comment by RaimoNiskanen** [comment]
`gen_udp:send/N` with `Destination` are "new" API:s.  The original ones has got `Host` and `Port`.
When adding these new API:s to `gen_udp` we also allowed for `Port` to be an `atom()` specifying a service name.

That a host may be named "any" or "loopback" in a name service is, as you suspected, the reason that we can not interpret these host names as IP address aliases, and we solved that snag as you see in `gen_udp`.

It seems we never got around to adding the corresponding API:s to `gen_tcp` and `gen_sctp`.
Pull requests are welcome...

Maybe the type spec for `Address` in `gen_tcp:connect/3,4` should be changed to `inet:hostname() | inet:ip_address() | inet:local_address()` to get rid of this confusing overlap.

**Comment by mxxk** [comment]
Thanks for the thorough explanation, @RaimoNiskanen. Perhaps `gen_tcp:connect` can be extended to something like the following to match `gen_udp:send` somewhat:

```erl
% (For brevity, this example uses type declarations rather than function
%  specifications with constraints.)

-type address() :: inet:hostname() |
                   inet:ip_address() |
                   inet:local_address().

-type destination() :: {inet:ip_address(), inet:port_number()} |
                       inet:family_address().

-type port() :: inet:port_number() |
                atom().

% `connect/2`: new API
-spec connect(Destination :: destination(), Options) -> Retval.

% `connect/3`: old/new API
-spec connect(Address :: address(), Port :: port(), Options) -> Retval;
      connect(Destination :: destination(), Options, Timeout) -> Retval.

% `connect/4`: old API
-spec connect(Address :: address(), Port :: port(), Options, Timeout) -> Retval.
```

- As you suggested, the spec for `address()` drops `socket_address()` (and therefore `'any'` and `'loopback'`), and thus avoids the ambiguity.
- `destination()` is identical to how `Destination` is used in `gen_udp:connect/3,5`.
- In variants of `connect` which accept an `Address` and `Port`, `port()` allows an atom service name.
- (Haven't looked at `gen_sctp` yet; hoping to get some feedback on these `gen_tcp` types first!)

I appreciate the PR invitation, and I'd be open to help, though admittedly I'm very new to Erlang. 🙂

**Comment by RaimoNiskanen** [comment]
Yes, something like that, I think.  `Options` and `Timeout` should be type specified to avoid clauses with overlapping types.

**Comment by mxxk** [comment]
Right on; just edited the type specifications ^
